webrtc/: Add message framing to support half-close and reset of stream

mxinden opened the pull request:
_Note this pull request targets https://github.com/libp2p/specs/pull/412 and not the master branch._

The WebRTC browser APIs do not support half-closing nor resets of streams. This
commit defines a message framing schema to support this functionality on top of
the browser APIs.

This is a first proposal. This has not yet been implemented and will thus most likely change.

Input very much appreciated.

See https://github.com/libp2p/specs/pull/412#discussion_r932487837 and https://github.com/libp2p/specs/pull/412#issuecomment-1209246438 for past discussions on this topic.

mxinden committed:
webrtc/: Add message framing to support half-close and reset of stream

The WebRTC browser APIs do not support half-closing nor resets of streams. This
commit defines a message framing schema to support this functionality on top of
the browser APIs.

mxinden commented:
For those that want to play around with this, framing size test is here https://github.com/mxinden/rust-libp2p/commit/59617de20d12761c3d7fe882e4efc706b53102fa.

mxinden commented:
I implemented this proposal on top of https://github.com/libp2p/rust-libp2p/pull/2622. See https://github.com/mxinden/rust-libp2p/tree/webrtc-message-framing. (Happy path only.)

mxinden committed:
webrtc/: Reword protobuf comment from local endpoint to sender

mxinden commented:
JS implementation of this proposal https://github.com/little-bear-labs/js-libp2p-webrtc/pull/20 from @John-LittleBearLabs.

marten-seemann commented on webrtc/README.md:
```diff
@@ -260,21 +262,50 @@ be the same. On mismatch the final Noise handshake MUST fail.
+message Message {
+  enum Flag {
+    // The sender will no longer send messages.
+    CLOSE_WRITE = 0;
```
```suggestion
    FIN = 0;
```

mxinden replied:
Done in https://github.com/mxinden/specs/pull/1/commits/93df7e397e7ba79962ed7bdfbf6d4da40985011c.

marten-seemann commented on webrtc/README.md:
```diff
@@ -260,21 +262,50 @@ be the same. On mismatch the final Noise handshake MUST fail.
+    // The sender will no longer send messages.
+    CLOSE_WRITE = 0;
+    // The sender will no longer read messages.
+    CLOSE_READ = 1;
```
```suggestion
    STOP_SENDING = 1;
```

mxinden replied:
Done in https://github.com/mxinden/specs/pull/1/commits/93df7e397e7ba79962ed7bdfbf6d4da40985011c.

marten-seemann commented on webrtc/README.md:
```diff
@@ -260,21 +262,50 @@ be the same. On mismatch the final Noise handshake MUST fail.
+    // The sender will no longer read messages.
+    CLOSE_READ = 1;
+    // The local endpoint abruptly terminates the stream. The remote endpoint
+    // may discard any in-flight data.
```
Let's not mix this up with the read side of the stream. `RESET` **only** resets write side, NOT the read side.

mxinden replied:
Works for me. Updated. See https://github.com/mxinden/specs/pull/1/commits/865f4f2dea8872c8de301a16b59000ac4540f18d.

mxinden replied:
@John-LittleBearLabs note the above change in semantics. Please comment in case you have questions / concerns.

melekes commented on webrtc/README.md:
```diff
@@ -260,21 +262,50 @@ be the same. On mismatch the final Noise handshake MUST fail.
-  additional multiplexing.
+Following [Connection Security](#connection-security).
+
+The WebRTC browser APIs do not support half-closing nor resets of streams.
```
just so we're on the same page. when you're talking about reseting a stream, you're not talking about https://datatracker.ietf.org/doc/html/draft-ietf-rtcweb-data-channel-13#section-6.7 (webrtc signalling mechanism to tell remote that data channel was closed), right?

mxinden replied:
The above description is misleading. Thanks for the note. I pushed https://github.com/mxinden/specs/pull/1/commits/a60234cfb44fe060c8448d0de6e91958a57fedbc updating the section.

WebRTC's reset as well as the corresponding `RTCDataChannel.close()` reset both sending and receiving. We would like to support the former without the latter.

All that said, in the end I don't see this mechanism as particularly useful. I expect readers to read messages from the stream in order, and not do any buffering to the application above. Thus a receiver would receive the `RESET` after having passed all previous messages to the application already. Thus acting just as if it received a `FIN` flag, i.e. closing its read side, but not discarding any messages.

I don't have a strong opinion on adding `RESET`. I guess it doesn't hurt and might enable future optimizations.

The above is not to be generalized to the entire proposal. While I don't see much value in the `RESET` flag, there is value in being able to half-close a stream via the `FIN` and `STOP_SENDING` flag.

Does the above make sense @melekes? These are rather subtle semantics. Please comment in case my explanations and reasoning in the proposal don't make sense to you.

melekes replied:
Thanks for the clarification 👍 The proposed framing looks good to me. Regarding `RESET` - depending on the implementation there could be some internal buffering. This buffer could be erased once `RESET` is received. Not sure how useful it would be in practice though.

mxinden committed:
webrtc/: Adapt enum variants to QUIC RFC

mxinden committed:
webrtc/: Link QUIC RFC as reference

mxinden committed:
webrtc/: Redefine RESET semantics

Instead of resetting both ends of a stream, RESET only resets the sending side
of a stream. This is in line with the QUIC stream reset mechanism.

mxinden committed:
webrtc/: Rephrase missing mechanism in browser webrtc api

mxinden commented:
@marten-seemann thanks for the review. I addressed your comments. Do you want to take another look?

melekes reviewed (approved):
👍

mxinden commented:
Unless there are any objections, I will merge here end of the week.

thomaseizinger commented on webrtc/README.md:
```diff
@@ -260,21 +262,55 @@ be the same. On mismatch the final Noise handshake MUST fail.
+
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and send on the `RTCDataChannel` prefixed with the message length in
```
```suggestion
below and sent on the `RTCDataChannel` prefixed with the message length in
```

thomaseizinger commented on webrtc/README.md:
```diff
@@ -339,3 +375,18 @@ After [Connection Security](#connection-security):
+  (`(5 bytes + 2 bytes) / 40 bytes = 0.175`) but likely irrelevant.
+
+  Using Protobuf allows us to evolve the protocol in a backwards compatibile way
+  going forward. Using Protobuf is consisten with the many other libp2p
```
```suggestion
  going forward. Using Protobuf is consistent with the many other libp2p
```

thomaseizinger reviewed (approved):
Looks good to me. Found two typos :)

marten-seemann reviewed (commented):
Do we need to specify what happens when receiving a STOP_SENDING? If we want to go in parallel with RFC 9000 we'd need to say that you send a RESET.

mxinden committed:
Apply suggestions from code review

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

mxinden committed:
webrtc/: Require RESET after STOP_SENDING

mxinden commented:
I was not aware of this convention @marten-seemann. Thanks for pointing it out. I think it makes sense to be consistent with RFC 9000. https://github.com/mxinden/specs/pull/1/commits/31ac65d6a78f621babd8d6c2845881b1128db4ad includes the convention for our adaptation of the QUIC multiplexer.

mxinden committed:
webrtc/: Rename RESET to RESET_STREAM

For the sake of consistency with the QUIC RFC, rename RESET to RESET_STREAM.

mxinden committed:
Merge branch 'webrtc' of github.com:mxinden/specs into webrtc-msg-framing

mxinden committed:
webrtc/: Use message framing for Noise handshake already

Instead of executing the additional Noise handshake on a plain `RTCDataChannel`,
frame the Noise messages with the proposed message framing mechanism already.

Benefits:

- The additional Noise handshake can already make use of the proper closing
mechanism. Note that `RTCDataChannel.close` may drop any in-flight data. See
https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close.

- Simplifies implementations as there is only one way to write on an `RTCDataChannel`.

- Allows us to role out new versions of the Noise handshake in the future by
adding a version to the message framing Protobuf.

mxinden commented:
Bringing https://github.com/mxinden/specs/pull/1/commits/7a8ebc03c1b514922adfbec6e6ac756c0c892599 to everyone's attention here:

> webrtc/: Use message framing for Noise handshake already
> 
> Instead of executing the additional Noise handshake on a plain `RTCDataChannel`,
> frame the Noise messages with the proposed message framing mechanism already.
> 
> Benefits:
> 
> - The additional Noise handshake can already make use of the proper closing
> mechanism. Note that `RTCDataChannel.close` may drop any in-flight data. See
> https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close.
> 
> - Simplifies implementations as there is only one way to write on an `RTCDataChannel`.
> 
> - Allows us to role out new versions of the Noise handshake in the future by
> adding a version to the message framing Protobuf.

Let me know in case you see any issues with this.

thomaseizinger commented on webrtc/README.md:
```diff
@@ -117,9 +116,11 @@ fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+   mechanism described in [Multiplexing](#multiplexing).
 
-8. See [Multiplexing](#multiplexing).
+7. The remote is authenticated via an additional Noise handshake. See
```
```suggestion
8. The remote is authenticated via an additional Noise handshake. See
```

thomaseizinger reviewed (approved):
LGTM!

mxinden committed:
Update webrtc/README.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

mxinden commented:
I am going to merge here. In case you have any follow-ups please raise them on https://github.com/libp2p/specs/pull/412.

mxinden commented:
_Moving a question from @melekes on https://github.com/melekes/rust-libp2p/pull/8#issuecomment-1252188039 here._

> > Send RESET_STREAM when receiving STOP_SENDING
> 
> again, not sure why it's needed. The remote has already indicated it won't accept any more data, so why send `RESET_STREAM`?

My rational for introducing it in this pull request was for the sake of consistency with the QUIC specification. I am assuming this can be useful for applications leveraging the acknowledgement of the closing of the read side. I can not find any past discussions on the matter

@marten-seemann do you know more?

marten-seemann commented:
> My rational for introducing it in this pull request was for the sake of consistency with the QUIC specification. I am assuming this can be useful for applications leveraging the acknowledgement of the closing of the read side. I can not find any past discussions on the matter
>
> @marten-seemann do you know more?

In QUIC it's needed to have an accurate accounting of the number of bytes sent for connection-level flow control purposes (RESET_STREAM contains the final offset of that stream). How does flow control work in WebRTC? Is it purely based on streams? Then we probably don't need to send a RESET_STREAM.

melekes commented:
> How does flow control work in WebRTC? Is it purely based on streams? Then we probably don't need to send a RESET_STREAM.

Each data channel is a SCTP stream (http://www.myreadingroom.co.in/notes-and-studymaterial/68-dcn/858-flow-control-in-sctp.html).

mxinden commented:
Thanks @marten-seemann and @melekes. Removed requirement with https://github.com/libp2p/specs/pull/412/commits/65a894a0d2c57b37f639f4d68ae28d9ea03be481.

//CC @ckousik and @John-LittleBearLabs
